aten/src/ATen/core/autocast/register_autocast_ops.cpp

```diff
@@ -0,0 +1,528 @@
+#include <ATen/core/op_registration/op_registration.h>
+#include <ATen/ATen.h>
+#include <ATen/NativeFunctions.h>
+// TODO:  Remove this dependency once callBoxed is possible for all ops
+#include <torch/csrc/jit/operator.h>
+
+#include <iostream>
+#include <exception>
+
+namespace at {
+namespace autocast {
+
+namespace {
+// Imitate Apex and cache some of the casts to streamline parameter reuse.
+// Our heuristic is to cache FP16 casts of FP32 model weights (see cached_cast below).
+//
+// After discussion with @ezyang, the cache uses the following structure:
+// The key is the source tensor's TensorImpl*, a proxy for a Tensor uuid that's unchanged across shallow copies.
+// The value is a tuple of two Tensors:  the source tensor and the casted tensor.  Both are stored as full
+// Tensors to ensure their TensorImpls stay alive if other references to them die.
+//
+// We must keep the source tensor alive because if the source tensor were deallocated, another random Tensor
+// could be allocated whose TensorImpl* happened to have the same value.  This TensorImpl* would then mistakenly
+// hit in cache, which would be a nasty bug (rare, intermittent, unpredictable).
```

Comment:
Note that you could have stored a weak reference to the source tensor inside the map, which would also solve this problem. (I don't recall if there was another reason to store a strong reference, I do recall we did talk about it)

Reply:
We did it because we discovered to our dismay that weak references were gone.  Maybe they've been added back since?

Reply:
intrusive_ptr still has weak references, and TensorImpl is implemented using intrusive ptr...

Reply:
Changed to save a weak_intrusive_ptr to the source's TensorImpl in the map.